### DISCUSSION TITLE
Linting with fypp questions

### DISCUSSION THREAD

**Discussion by jphaupt** [krvajal/vscode-fortran-support#717] [Q&A - deprecated see Fortran Discourse]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

I noticed fypp support has recently been added, thanks a lot!! I understand it's still new, but I've noticed a little bug: 
- First (I suspect this is something to do with my configuration?) if my file extension is `.fpp` then it looks like nothing is recognised, there are red squiggles everywhere (I have it configured for Fortran). If I rename it to have `.fypp` extension and don't change anything else, suddenly it's fine. 
- If I have a derived type that uses the preprocessor, then any code that comes after that derived type loses all syntax highlighting. 

In the attached screenshot, all the highlighting goes back the way I expect it to be if I: 
- rename to `.fypp`
- remove the derived type

### Screenshots

![image](https://user-images.githubusercontent.com/10662835/196241533-1d66d430-2744-42c7-8437-5b936e18f9ea.png)

### Code Example

```Fortran
program highlighting_eg
    use iso_fortran_env, only: sp => real32, dp => real64
    implicit none
    #:for dtype in ['real', 'complex']
    #:for prec in ['sp', 'dp']
    ${dtype}$(${prec}$) :: a_${dtype}$_${prec}$
    #:endfor
    #:endfor

    #:for dtype in ['real', 'complex']
    type literally_anything_${dtype}$_t
    endtype literally_anything_${dtype}$_t
    #:endfor

    contains

    subroutine print(arg)
        real, intent(in) :: arg
        print*, arg
    end subroutine print
end program highlighting_eg
```

### Fortran Version (Free/Fixed Form)

Free Form e.g. F90+

### Version of Modern Fortran

v3.2.0

### Version of Visual Studio Code

v1.72.2

### Platform and Architecture

Linux (OpenSUSE)

### Additional Information

I noticed my other issue got sent into discussion. Should this also be there?

**Comment by gnikit** [comment]
So the main thing you noticed is a Duplicate issue of https://github.com/fortran-lang/vscode-fortran-support/issues/617. I will not close this just yet as a duplicate because I want to check what you mentioned about the extensions. That is strange indeed.

**Comment by jphaupt** [comment]
> So the main thing you noticed is a Duplicate issue of #617. I will not close this just yet as a duplicate because I want to check what you mentioned about the extensions. That is strange indeed.

Ah sorry, missed the type definition in that issue. Anyway, the extension issue is easy to work around, though a little inconvenient.

**Comment by gnikit** [comment]
I am not able to reproduce the behaviour you pointed out. I was able to get fypp diagnostics for the .fpp file. Is there something else I should be doing to recreate your issue?

![image](https://user-images.githubusercontent.com/16143716/196247373-9ac50ffa-a164-459d-b4cb-56d7e1f5aecf.png)

**Reply by jphaupt** [reply]
No, any file I open with .fpp as an extension has this issue, I don't do anything else. Weird. I'll play around with my settings settings to see if I can find the culprit.

**Reply by jphaupt** [reply]
I just noticed that you get full fypp diagnostics whereas even with .fypp I get a linker warning. I removed all settings (both local and global, exception: `"fortran.linter.modOutput": "${workspaceFolder}/.fmod"`) and disabled all other extensions (also updated VSCode and also tried Insiders and the pre-release extension). Here is a screenshot with the two extensions side by side: 

![image](https://user-images.githubusercontent.com/10662835/196374350-cd2ca909-e53e-4b07-a0dc-d289632f73a1.png)

Unfortunately searching for the warning hasn't been very helpful.

**Reply by jphaupt** [reply]
Of course as soon as I post about it I get that sorted. `"fortran.linter.fypp.enabled"` was false by default, making it true fixes it, although I still don't get any diagnostics (I just get no warnings/errors), both for fypp and fpp. Edit: Looks like that doesn't do what I want either because it messes up linting in a bigger project. When I have time I'll try to reproduce it in a minimal example, but it is the middle of a workday right now...

**Reply by gnikit** [reply]
I was about to suggest to check if linting for fypp was on. As you noticed `fypp` is support is still in its early stages. To prevent the creation of temporary files, which is expensive, Modern Fortran currently does everything though IO pipe streams. This is an undocumented "feature" of gfortran so it's very likely that things will break. Also, the fact that we define all linter options globally has probably something to do with it.

**Comment by gnikit** [comment]
Converted into a discussion since I could not find a bug or replicate the linting. Feel free to comment with more details to sort this out.

**Reply by gnikit** [reply]
PS I changed the title to the linting issue
